### PR TITLE
Postpone create parquet file handle until really needed

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

Pospone create parquet file handle until really needed.
Also use index statistics to get totalRows

## How was this patch tested?

current test pass. performance verified by Baidu.

### CODE REVIEW

**Comment by LuciferYang** [comment]
@linhong-intel @xuanyuanking It's  really a problem that we can't get totalRows if we don't actually read parquet data file, but at the moment I think performance is important than this metrics value because read footer every task may loss three times performance. We should find ways to get totalRows on the premise of ensuring performance, such as obtaining totalRows from the index statistics.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 1]
```diff
@@ -283,31 +283,28 @@ private[sql] class OapFileFormat extends FileFormat
...
+          if (m.dataReaderClassName == OapFileFormat.OAP_DATA_FILE_CLASSNAME) {
+            val dataFile = DataFile(file.filePath, m.schema, m.dataReaderClassName, conf)
+            val dataFileHandle: OapDataFileHandle = DataFileHandleCacheManager(dataFile)
+            if (filters.exists(filter =>
+              canSkipFile(dataFileHandle.columnsMeta.map(_.statistics), filter, m.schema))) {
+              val totalRows = dataFileHandle.totalRowCount()
+              selectedRows.add(0)
+              skippedRows.add(totalRows)
+              skip = true
+            }
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: Should we extract this part to a new method return Boolean value and we can avoid def a var ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 2]
```diff
@@ -193,6 +193,8 @@ private[oap] class OapDataReader(
 
   var selectedRows: Option[Long] = None
 
+  def totalRows(): Long = 0
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: Can we get totalRows value from the index statistics now ?

Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: Yes, I'm working on this today. no worry.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 1 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: good idea

**Comment by LuciferYang** [comment]
@linhong-intel Next Monday I will test the impact of this patch on performance

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 3]
```diff
@@ -381,6 +384,10 @@ private[oap] case class BitMapScanner(idxMeta: IndexMeta) extends IndexScanner(i
     bmUniqueKeyListCount = 0
   }
 
+  // Set by analyzeStatistics()
+  private var _totalRows: Long = 0
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: should we reset _totalRows value to 0 in clearCache method , and move this line to line 53 to put vars declarations together?

**Comment by adrian-wang** [comment]
LGTM, can you also create a pr against master branch?

**Comment by linhong-intel** [comment]
Thanks, @adrian-wang I'll take care of master branch.

**Pull request merged**